hs_core/hs_rdf.py

```diff
@@ -0,0 +1,358 @@
+from rdflib.collection import Collection
+from rdflib.namespace import Namespace, NamespaceManager, DC, DCTERMS, RDF, RDFS
+from rdflib.plugin import register
+from rdflib.plugins.serializers.rdfxml import XMLLANG, OWL_NS, XMLBASE
+from rdflib.plugins.serializers.xmlwriter import XMLWriter
+from rdflib.serializer import Serializer
+from rdflib.term import Literal, URIRef
+from rdflib.util import first
+
+HSTERMS = Namespace("http://hydroshare.org/terms/")
+RDFS1 = Namespace("http://www.w3.org/2000/01/rdf-schema#")
+
+NAMESPACE_MANAGER = NamespaceManager(Graph())
+NAMESPACE_MANAGER.bind('hsterms', HSTERMS, override=False)
+NAMESPACE_MANAGER.bind("rdfs1", RDFS1, override=False)
+NAMESPACE_MANAGER.bind('dc', DC, override=False)
+NAMESPACE_MANAGER.bind('dcterms', DCTERMS, override=False)
+
+
+class RDF_MetaData_Mixin(object):
+    """
+    A mixin for MetaData objects which store their metadata in generic relations.  If metadata outside of generic
+    relations need to be used, you may extend ingest_metadata and get_rdf_graph to include the other metadata elements
+    """
+
+    def rdf_subject(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_subject")
+
+    def rdf_metadata_subject(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_metadata_subject")
+
+    def rdf_type(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_type")
+
+    @classmethod
+    def rdf_subject_from_graph(cls, graph):
+        """Derive the root subject of an rdflib Graph by returning the subject in the triple with predicate DC.title"""
+        subject = None
+        for s, _, _ in graph.triples((None, DC.title, None)):
+            subject = s
+            break
+        if not subject:
+            raise Exception("Invalid rdf/xml, could not find required predicate dc:title")
+        return subject
+
+    def ingest_metadata(self, graph):
+        """Given an rdflib Graph, run ingest_rdf for all generic relations on the object"""
+        subject = self.rdf_subject_from_graph(graph)
+
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            generic_relation.related_model.ingest_rdf(graph, subject, self)
+        self.save()
+
+    def get_rdf_graph(self):
+        """adds the rdf triples of all generic relations on the object into an rdflib Graph"""
+        graph = Graph()
+        graph.namespace_manager = NAMESPACE_MANAGER
+
+        subject = self.rdf_subject()
+        # graph.add((subject, RDF.type, self.rdf_type()))
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            for f in getattr(self, getattr(generic_relation, 'name', None), None).all():
```

Comment:
This for statement may be problematic since if attr does not exist, it defaults to None, and ```None.all()``` will raise an exception. Also, commented line 64 can be removed.

Reply:
the generic relation will always have a name and self will always have an attribute with the generic relation name... so None will never be returned in either getattr.  But it is confusing, so I'll go ahead and break up the statement.  How's this below?
```
gr_name = generic_relation.name
gr = getattr(self, gr_name)
for f in gr.all():
'''

Reply:
Yeah, the break up logic is clearer. Alternatively, you can just remove the default ```None``` from the original getattr() calls since the default None gave the indication that could occur which could cause confusion.